--- linux.orig/arch/mips/mm/c-r4k.c	2004-07-13 08:38:53.000000000 +0900
+++ linux/arch/mips/mm/c-r4k.c	2004-09-02 00:05:59.000000000 +0900
@@ -736,6 +736,7 @@
 	case CPU_VR4133:
 		write_c0_config(config & ~CONF_EB);
 	case CPU_VR4131:
+#if 0
 		/* Workaround for cache instruction bug of VR4131 */
 		if (c->processor_id == 0x0c80U || c->processor_id == 0x0c81U ||
 		    c->processor_id == 0x0c82U) {
@@ -743,6 +744,7 @@
 			config |= 0x00410000U;
 			write_c0_config(config);
 		}
+#endif
 		icache_size = 1 << (10 + ((config & CONF_IC) >> 9));
 		c->icache.linesz = 16 << ((config & CONF_IB) >> 5);
 		c->icache.ways = 2;